Clients can crash server by tampering with preferred color

- Intercept other players preferred color any way you like
- Put it to your settings.yaml
- Join lobby with that player
- Click on empty slot (change your slot), your color will become your preferred color (it's now two players with the same color in lobby)

After this if new client will try to connect to the server it will connect and then immediately disconnect.
If admin will try to add new bots server will crash with exception:
```
Exception of type `System.ArgumentException`: An element with the same key already exists in the dictionary.
  at System.Collections.Generic.Dictionary`2[System.Drawing.Color,System.String].Add (Color key, System.String value) [0x00000] in <filename unknown>:0 
  at System.Linq.Enumerable.ToDictionary[Client,Color,String] (IEnumerable`1 source, System.Func`2 keySelector, System.Func`2 elementSelector, IEqualityComparer`1 comparer) [0x00000] in <filename unknown>:0 
  at System.Linq.Enumerable.ToDictionary[Client,Color,String] (IEnumerable`1 source, System.Func`2 keySelector, System.Func`2 elementSelector) [0x00000] in <filename unknown>:0 
  at OpenRA.Mods.Common.Server.ColorValidator.ValidatePlayerNewColor (OpenRA.Server.Server server, Color askedColor, Int32 playerIndex, System.Drawing.Color& forbiddenColor, OpenRA.Server.Connection connectionToEcho) [0x0007f] in OpenRA.Mods.Common/ServerTraits/ColorValidator.cs:163 
  at OpenRA.Mods.Common.Server.ColorValidator.ValidatePlayerNewColor (OpenRA.Server.Server server, Color askedColor, Int32 playerIndex, OpenRA.Server.Connection connectionToEcho) [0x00000] in OpenRA.Mods.Common/ServerTraits/ColorValidator.cs:198 
  at OpenRA.Mods.Common.Server.LobbyCommands+<InterpretCommand>c__AnonStorey68.<>m__17 (System.String s) [0x001af] in OpenRA.Mods.Common/ServerTraits/LobbyCommands.cs:287 
  at OpenRA.Mods.Common.Server.LobbyCommands.InterpretCommand (OpenRA.Server.Server server, OpenRA.Server.Connection conn, OpenRA.Network.Client client, System.String cmd) [0x00313] in OpenRA.Mods.Common/ServerTraits/LobbyCommands.cs:870 
  at OpenRA.Server.Server.InterpretServerOrder (OpenRA.Server.Connection conn, OpenRA.Server.ServerOrder so) [0x000bb] in OpenRA.Game/Server/Server.cs:462 
  at OpenRA.Server.Server.InterpretServerOrders (OpenRA.Server.Connection conn, System.Byte[] data) [0x00025] in OpenRA.Game/Server/Server.cs:447 
  at OpenRA.Server.Server.DispatchOrders (OpenRA.Server.Connection conn, Int32 frame, System.Byte[] data) [0x0000c] in OpenRA.Game/Server/Server.cs:416 
  at OpenRA.Server.Connection.ReadData (OpenRA.Server.Server server) [0x000b4] in OpenRA.Game/Server/Connection.cs:106 
  at OpenRA.Server.Server.<Server>m__1D (System.Object _) [0x001c2] in OpenRA.Game/Server/Server.cs:180 
  at System.Threading.Thread.StartInternal () [0x00000] in <filename unknown>:0 
[ERROR] FATAL UNHANDLED EXCEPTION: System.ArgumentException: An element with the same key already exists in the dictionary.
  at System.Collections.Generic.Dictionary`2[System.Drawing.Color,System.String].Add (Color key, System.String value) [0x00000] in <filename unknown>:0 
  at System.Linq.Enumerable.ToDictionary[Client,Color,String] (IEnumerable`1 source, System.Func`2 keySelector, System.Func`2 elementSelector, IEqualityComparer`1 comparer) [0x00000] in <filename unknown>:0 
  at System.Linq.Enumerable.ToDictionary[Client,Color,String] (IEnumerable`1 source, System.Func`2 keySelector, System.Func`2 elementSelector) [0x00000] in <filename unknown>:0 
  at OpenRA.Mods.Common.Server.ColorValidator.ValidatePlayerNewColor (OpenRA.Server.Server server, Color askedColor, Int32 playerIndex, System.Drawing.Color& forbiddenColor, OpenRA.Server.Connection connectionToEcho) [0x0007f] in OpenRA.Mods.Common/ServerTraits/ColorValidator.cs:163 
  at OpenRA.Mods.Common.Server.ColorValidator.ValidatePlayerNewColor (OpenRA.Server.Server server, Color askedColor, Int32 playerIndex, OpenRA.Server.Connection connectionToEcho) [0x00000] in OpenRA.Mods.Common/ServerTraits/ColorValidator.cs:198 
  at OpenRA.Mods.Common.Server.LobbyCommands+<InterpretCommand>c__AnonStorey68.<>m__17 (System.String s) [0x001af] in OpenRA.Mods.Common/ServerTraits/LobbyCommands.cs:287 
  at OpenRA.Mods.Common.Server.LobbyCommands.InterpretCommand (OpenRA.Server.Server server, OpenRA.Server.Connection conn, OpenRA.Network.Client client, System.String cmd) [0x00313] in OpenRA.Mods.Common/ServerTraits/LobbyCommands.cs:870 
  at OpenRA.Server.Server.InterpretServerOrder (OpenRA.Server.Connection conn, OpenRA.Server.ServerOrder so) [0x000bb] in OpenRA.Game/Server/Server.cs:462 
  at OpenRA.Server.Server.InterpretServerOrders (OpenRA.Server.Connection conn, System.Byte[] data) [0x00025] in OpenRA.Game/Server/Server.cs:447 
  at OpenRA.Server.Server.DispatchOrders (OpenRA.Server.Connection conn, Int32 frame, System.Byte[] data) [0x0000c] in OpenRA.Game/Server/Server.cs:416 
  at OpenRA.Server.Connection.ReadData (OpenRA.Server.Server server) [0x000b4] in OpenRA.Game/Server/Connection.cs:106 
  at OpenRA.Server.Server.<Server>m__1D (System.Object _) [0x001c2] in OpenRA.Game/Server/Server.cs:180 
  at System.Threading.Thread.StartInternal () [0x00000] in <filename unknown>:0 
```

<bountysource-plugin>
---
Want to back this issue? **[Post a bounty on it!](https://www.bountysource.com/issues/12433203-clients-can-crash-server-by-tampering-with-preferred-color?utm_campaign=plugin&utm_content=tracker%2F36085&utm_medium=issues&utm_source=github)** We accept bounties via [Bountysource](https://www.bountysource.com/?utm_campaign=plugin&utm_content=tracker%2F36085&utm_medium=issues&utm_source=github).
</bountysource-plugin>

--- a/OpenRA.Mods.Common/ServerTraits/ColorValidator.cs
+++ b/OpenRA.Mods.Common/ServerTraits/ColorValidator.cs
@@ -160,17 +160,11 @@ namespace OpenRA.Mods.Common.Server
 			}
 
 			// Validate color against other clients
-			var playerColors = server.LobbyInfo.Clients
-				.Where(c => c.Index != playerIndex)
-				.ToDictionary(c => c.Color.RGB, c => c.Name);
-
-			if (!ValidateColorAgainstForbidden(askedColor, playerColors.Keys, out forbiddenColor))
+			var playerColors = server.LobbyInfo.Clients.Where(c => c.Index != playerIndex).Select(c => c.Color.RGB);
+			if (!ValidateColorAgainstForbidden(askedColor, playerColors, out forbiddenColor))
 			{
 				if (connectionToEcho != null)
-				{
-					var client = playerColors[forbiddenColor];
-					server.SendOrderTo(connectionToEcho, "Message", "Color was too similar to {0}, and has been adjusted.".F(client));
-				}
+					server.SendOrderTo(connectionToEcho, "Message", "Color was too similar to another player's color, and has been adjusted.");
 
 				return false;
 			}

